### ISSUE TITLE
VOTable writer does not write out GROUPs within RESOURCEs.

### ISSUE THREAD

**Issue by tomdonaldson** [astropy/astropy#16511]
### Description

When writing out a VOTable from `VOTableFile` object, any `Group` children of `Resource` objects are not written out.

`VOTableFile`, `TableElement` and `Group` objects can also have children of type `Group`, and in all those cases the `Group` children are properly written out.

### Expected behavior

If I read in a VOTable containing a `RESOURCE` element that has a child `GROUP` element, or I construct such a `VOTableFile` internally using the `astropy.io.votable` API, then I write out the VOTable to xml, I expect the `GROUP` child of the `RESOURCE` to be included in the output.

### How to Reproduce

This script reads in `test_in.xml` which has `GROUP` children for `VOTABLE`, `GROUP`, `RESOURCE` and `TABLE`.

Then it prints the contents of the `GROUP` under the `RESOURCE` to show that it has not been lost on read.

Then it writes the VOTable to `test_out.xml`.  Note that the `GROUP` under the `RESOURCE` is missing.

```python
from astropy.io.votable import parse

votable = parse("test_in.xml")
print(f"{votable.resources[0].groups[0]=}")
votable.to_xml("test_out.xml")
```

Output:
```
votable.resources[0].groups[0]=<GROUP>... 1 entries ...</GROUP>
```

test_in.xml:
```xml
<?xml version="1.0" encoding="utf-8"?>
<VOTABLE version="1.4" xmlns="http://www.ivoa.net/xml/VOTable/v1.3" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.ivoa.net/xml/VOTable/v1.3 http://www.ivoa.net/xml/VOTable/VOTable-1.4.xsd">
     <GROUP ID="VOTABLE_child">
          <PARAM arraysize="*" datatype="char" name="VOTABLE_group_param" value=""/>
               <GROUP ID="GROUP_child">
                    <PARAM arraysize="*" datatype="char" name="GROUP_group_param" value=""/>
               </GROUP>
     </GROUP>
     <RESOURCE ID="soda-service" type="meta" utype="adhoc:service">
          <PARAM ID="standardID" arraysize="*" datatype="char" name="standardID" value="ivo://ivoa.net/std/SODAnext#sync-1.0"/>
          <PARAM ID="accessURL" arraysize="*" datatype="char" name="accessURL" value="https://abc.net"/>
          <GROUP ID="RESOURCE_child">
               <PARAM arraysize="*" datatype="char" name="RESOURCE_group_param" value=""/>
          </GROUP>
          <TABLE>
               <GROUP ID="TABLE_child">
                    <PARAM arraysize="*" datatype="char" name="TABLE_group_param" value=""/>
               </GROUP>
          </TABLE>
     </RESOURCE>
</VOTABLE>
```

test_out.xml:
```xml
<?xml version="1.0" encoding="utf-8"?>
<!-- Produced with astropy.io.votable version 6.1.0
     http://www.astropy.org/ -->
<VOTABLE version="1.4" xmlns="http://www.ivoa.net/xml/VOTable/v1.3" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.ivoa.net/xml/VOTable/v1.3 http://www.ivoa.net/xml/VOTable/VOTable-1.4.xsd">
 <GROUP ID="VOTABLE_child">
  <PARAM ID="VOTABLE_group_param" arraysize="*" datatype="char" name="VOTABLE_group_param" value=""/>
  <GROUP ID="GROUP_child">
   <PARAM ID="GROUP_group_param" arraysize="*" datatype="char" name="GROUP_group_param" value=""/>
  </GROUP>
 </GROUP>
 <RESOURCE ID="soda-service" type="meta" utype="adhoc:service">
  
  <PARAM ID="standardID" arraysize="*" datatype="char" name="standardID" value="ivo://ivoa.net/std/SODAnext#sync-1.0"/>
  <PARAM ID="accessURL" arraysize="*" datatype="char" name="accessURL" value="https://abc.net"/>
  <TABLE>
   <GROUP ID="TABLE_child">
    <PARAM ID="TABLE_group_param" arraysize="*" datatype="char" name="TABLE_group_param" value=""/>
   </GROUP>
  </TABLE>
 </RESOURCE>
</VOTABLE>
```

### Versions

```python
import astropy
try:
    astropy.system_info()
except AttributeError:
    import platform; print(platform.platform())
    import sys; print("Python", sys.version)
    import astropy; print("astropy", astropy.__version__)
    import numpy; print("Numpy", numpy.__version__)
    import erfa; print("pyerfa", erfa.__version__)
    try:
        import scipy
        print("Scipy", scipy.__version__)
    except ImportError:
        print("Scipy not installed")
    try:
        import matplotlib
        print("Matplotlib", matplotlib.__version__)
    except ImportError:
        print("Matplotlib not installed")
```
```
macOS-14.4.1-arm64-arm-64bit
Python 3.10.14 (main, May  6 2024, 14:42:37) [Clang 14.0.6 ]
astropy 6.1.0
Numpy 1.26.4
pyerfa 2.0.1.4
Scipy not installed
Matplotlib not installed
```

**Comment by tomdonaldson** [comment]
It's very possible that there are other children of various elements that are lost in a round trip like this.  Ideally a roundtrip from xml to `VOTableFile` object back to xml should preserve all the content.  Preserving all that content is more challenging when going all the way to an Astropy `Table`, so that is out of scope for this issue.

**Comment by pllim** [comment]
Is this related to https://github.com/astropy/astropy/issues/8705 ?

**Comment by tomdonaldson** [comment]
> Is this related to #8705 ?

Yes, I see that this issue is the same a comment in #8705: https://github.com/astropy/astropy/issues/8705#issuecomment-546580339

#8705 mainly seems to be addressing the lack of a clear public API for constructing the contents of a `Group` object.  It can be done by adding the `Group.entries` list (instead of the more intuitive `Group.params`, `Group.groups`, etc.), but that's not documented.  I wonder if that capability means we can close #8705, or whether it is asking for something better.  Without fixing this issue, it's almost moot whether one can create the Group or not.

Programmatically adding children to many of those objects is sloppy already, as discussed in #16314 (specifically about `fields`) since the user is expected to just add to an existing list of children instead of calling some sort of method (`add_some_child()`) that wraps that functionality.

**Comment by pllim** [comment]
Thanks for the clarification. In that case, maybe I'll mark that issue as being blocked by this one.

**Comment by bsipocz** [comment]
Your example above is not exactly round tripping as I also see new IDs within the PARAMS. Is that expected/OK?

**Comment by bsipocz** [comment]
@tomdonaldson - what is your updated take here, can we close https://github.com/astropy/astropy/issues/8705
